### PR TITLE
Refactor Tool Panel views structures and combine ToolBox and ToolBoxWorkflow into one component

### LINKED ISSUE #16101: Refactor Tool Panel views into one Toolbox
Currently, we have 3 toolbox views in the `panelStore`: default, EDAM ontology and EDAM topics:
- api/tools?in_panel=true&view=default
- api/tools?in_panel=true&view=ontology:edam_topics
- api/tools?in_panel=true&view=ontology:edam_operations

The only thing that changes in those three is the ordering of sections; the data for each tool remains the same.

Instead of having three views in the client store, we should only have one central flattened toolbox (`api/tools?in_panel=false`) with each tool having sufficient information about sections/ontologies, **so that this information can be accessed at all times, regardless of what "view" is currently enabled**.

In addition to the `api/tools?in_panel=false`, we would then GET 3 views from the backend, each returning the `ToolSection` details, with `section.elems` not being a list of `Tool`s and instead a list of `tool_id`s from each section. _This would mean a change to the backend tool structure..._

xref https://github.com/galaxyproject/galaxy/issues/16453#issuecomment-1649341706
Marius has outlined the idea here ^

Once we have the 3 section views, and one flattened toolbox from the backend, we would then create 3 front-end views for the ToolPanel.

### LINKED ISSUE #16499: Refactor ToolBox and ToolBoxWorkflow into one component
There is duplication in both components and it would make sense to have a single component for the `ToolBox`

### PR DESCRIPTION
**Pull request by ahmedhamidawan**
Changed the structure of toolbox received from the backend ToolBox is now an object of ToolSections (and Tools) by id
(sample structure laid out here in https://github.com/galaxyproject/galaxy/issues/16453#issuecomment-1649341706

### `api/tools?in_panel=True&view=ontology:edam_operations` used to return:
```
[
  {
    "model_class": "ToolSection",
    "id": "operation_0292",
    "name": "Sequence alignment",
    "elems": [ { Tool object }, { Tool object }, ...]
  },
  {
    "model_class": "ToolSection",
    "id": "operation_0491",
    "name": "Pairwise sequence alignment",
    "elems": [ { Tool object }, { Tool object }, ...]
  },
  ...
]
```

### and it now returns:
```
{
  "ontology:edam_operations": {
      "operation_0292": {
        "model_class": "ToolSection",
        "id": "operation_0292",
        "name": "Sequence alignment",
        "tools": ["tool_id", "tool_id", ...],
      },
      "operation_0491": {
        "model_class": "ToolSection",
        "id": "operation_0491",
        "name": "Pairwise sequence alignment",
        "tools": ["tool_id", "tool_id", ...],
      },
      ...
  }
}
```

And we use `api/tools?in_panel=false` to return a `toolsById` object to reference these views.

Going forward, this will make it easier for us to make any structural changes with these lighter "view" objects that store only tool ids, to then create the Tool Panel views we like at the front end.

### Note: Case where a `ToolSection` contains a `ToolSectionLabel`
In such cases (a "_Sub_" Label within a Section):
image
the new api structure will do something like this:
```
"filter": {
  "model_class": "ToolSection",
  "id": "filter",
  "name": "Filter and Sort",
  "version": "",
  "description": null,
  "links": null,
  "tools": [
    "Filter1",
    "sort1",
    "Grep1",
    {
      "model_class": "ToolSectionLabel",
      "id": "gff",
      "text": "GFF",
      "version": "",
      "description": null,
      "links": null
    },
    "Extract_features1",
    "gff_filter_by_attribute",
    "gff_filter_by_feature_count",
    "gtf_filter_by_attribute_values_list"
  ]
},
```

Fixes #16499 
Fixes #16101

Also consolidated `ToolBox` and `ToolBoxWorkflow` into one `ToolBox` that is contained within a `ToolPanel`. Removed Tool panel providers and ProviderAware tool boxes

## How to test the changes?
(Select all options that apply)
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [x] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

### CODE REVIEW

File: lib/galaxy/tool_util/toolbox/base.py [hunk 1]
```diff
@@ -1312,6 +1312,39 @@ def to_dict(self, trans, in_panel=True, tool_help=False, view=None, **kwds):
...
+        For in_panel=False, uses the structure:
+            {tools: {tool_id: tool.to_dict(), ...}}}
+        """
+        rval = {}
+        if in_panel:
+            if view is None:
+                view = self._default_panel_view
+            view_contents = {}
+            rval[view] = view_contents
+            panel_elts = list(self.tool_panel_contents(trans, view=view, **kwds))
```
Comment (mvdbeek) [inline lib/galaxy/tool_util/toolbox/base.py]:
```suggestion
            panel_elts = self.tool_panel_contents(trans, view=view, **kwds)
```

**Review by mvdbeek — COMMENTED** [review]
Backend looks good so far!

**Comment by jmchilton** [comment]
Parts of this seem correct. I certainly like the idea of just sending the backbone of tool sections - that seems like such a conceptual win from an API perspective. We're still implementing multiple different tool panel views and of different types and such though right? The thing that causes some anxiety on my end is we had like turned ontologies into just another tool panel view and now we're like recomputing some ontology code on the client that had shifted to the backend. It might be used in totally fine ways though. Cool stuff though - even if it makes me anxious!

**Comment by ahmedhamidawan** [comment]
<details>
  <summary> John's comment... </summary>

> Parts of this seem correct. I certainly like the idea of just sending the backbone of tool sections - that seems like such a conceptual win from an API perspective. We're still implementing multiple different tool panel views and of different types and such though right? The thing that causes some anxiety on my end is we had like turned ontologies into just another tool panel view and now we're like recomputing some ontology code on the client that had shifted to the backend. It might be used in totally fine ways though. Cool stuff though - even if it makes me anxious!

</details>

Thank you @jmchilton !

> We're still implementing multiple different tool panel views and of different types and such though right?

Yes, we're still treating those as the same (i.e.: At the front end, the Tool panel itself appears and works the exact same way.
Only difference is that the structure received from the backend has changed (as i have updated in the comment above

**Comment by jmchilton** [comment]
Thanks for the clarification - I appreciate you easing my anxieties! This is exciting work - very nice job.

**Comment by ElectronicBlueberry** [comment]
Regarding this example:

```json
"filter": {
  "model_class": "ToolSection",
  "id": "filter",
  "name": "Filter and Sort",
  "version": "",
  "description": null,
  "links": null,
  "tools": [
    "Filter1",
    "sort1",
    "Grep1",
    "panel_label_gff",       <- this part
    "Extract_features1",
    "gff_filter_by_attribute",
    "gff_filter_by_feature_count",
    "gtf_filter_by_attribute_values_list"
  ],
  "panel_labels": [      <- and this part
    {
      "model_class": "ToolSectionLabel",
      "id": "gff",
      "text": "GFF",
      "version": "",
      "description": null,
      "links": null
    }
  ]
},
```

If I understand this correctly, this treats strings within the "tools" array which are prefixed with "panel_label" in a special manner, referencing a label defined in "panel_labels".

If this is the case, I do not like the design choice of giving some string special type implications. What I mean by this is that all other strings here are tool identifiers, while "panel_label_gff" is a panel label identifier. This is confusing when working with the API return, and can lead to edge cases if a tool id happens to start with "panel_label_".

I'd prefer the label object to be directly inside the array, next to the IDs. This still gives the array a mixed type, which is not ideal, but now that mixed type is obvious and easier to handle, rather than be hidden behind indirection.

**Comment by ahmedhamidawan** [comment]
Can't figure why the jests are failing here, they are passing locally 🤔 

For the failing seleniums, like:
```
FAILED lib/galaxy_test/selenium/test_tool_form.py::TestToolForm::test_rerun_dataset_collection_element - selenium.common.exceptions.TimeoutException: Message: Timeout waiting on CSS selector [a[href$="tool_runner?tool_id=identifier_multiple"]] to become present.
```
they all relate to a timeout waiting on CSS selectors for the tool panel, for which the frontend structure is now changed; the `ToolBox` component is within the parent `ToolPanel` component, and if all tools are not loaded yet, we get a `LoadingSpan` in place of the `ToolBox`.
All these seleniums (the ones relating to this PR) are passing locally, since (somehow...) locally, there is no timeout fail and the selectors are found. It might have something to do with the `ToolBox` not loading in time here in the Github runs...?

**Review by dannon — APPROVED** [review]

**Comment by dannon** [comment]
@ahmedhamidawan and I talked about the API changes a bit and the plan is to adjust this so that the old route doesn't change and will continue to return the same to_dict, supporting `to_panel_view` instead via an extra query argument (or explicit route under `/api/tools/`?).

**Pull request merged**